README


This archive contains the distribution Dist-Zilla-Plugin-OptionalFeature,
version 0.020:

  Specify prerequisites for optional features in your distribution

This software is copyright (c) 2013 by Karen Etheridge.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.025.



Changes
Revision history for Dist-Zilla-Plugin-OptionalFeature

0.020     2014-11-29 04:53:50Z
          - when setting up prompting in Makefile.PL for multiple optional
            features, control all prompts from one master plugin to ensure the
            prompt order remains as desired

0.019     2014-11-22 07:35:51Z
          - make the -description option mandatory, for more readable prompt
            strings
          - prompt with just the description, as the feature name is usually
            very confusing for the user
          - handle funky characters in the feature description and other
            fields

0.018     2014-11-01 03:21:38Z
          - temporarily disable a test that relies on CPAN::Meta::Merge
            behaviour that does not yet exist

0.017     2014-09-19 06:16:11Z
          - new -prompt feature (enabled by default for 'required' feature
            prereqs), which interactively prompts the user about the feature

0.016     2014-09-06 22:39:49Z
          - now *properly* allowing unrecognized options (0.015)

0.015     2014-09-06 07:46:34Z
          - now allowing unrecognized options, with a warning, rather than
            dying (allows proactively releasing distributions using
            not-yet-implemented options)

0.014     2014-05-18 05:04:39Z
          - fix examples to use working code, that doesn't violate naming
            constraints for features (RT#94130)

0.013     2014-02-19 03:48:35Z
          - bump the prereq on JSON::Any, to ensure we know how to handle
            JSON::XS 3.x

0.012     2014-01-29 04:09:38Z
          - report more installed versions, as an attempt to debug some odd
            cpantesters failures

0.011     2014-01-21 04:10:42Z
          - new -require_develop option (which has always defaulted to true)

0.010     2014-01-18 22:01:56Z
          - fix config section in generated metadata that was not
            spec-compliant

0.009     2014-01-18 20:41:13Z
          - include our full config data in target dist's metadata

0.008     2013-11-30 17:26:03Z
          - re-release to fix compile test

0.007     2013-09-11 05:03:52Z
          - warnings tests bypassed during installation

0.006     2013-09-08 17:53:53Z
          - add test requirement on the minimal Config::MVP needed to handle
            an edge case that we test our handling of
          - new EXPERIMENTAL -default config option, instructing CPAN clients
            to install the features when running non-interactively

0.005     2013-09-02 23:53:12Z
          - no longer setting dynamic_config => 1 -- it is generally the CPAN
            client that interrogates the user, not Makefile.PL/Build.PL
          - detect invalid syntax in feature name (e.g. spaces are not
            permitted), in case the dist isn't validating its metadata syntax
            itself

0.004     2013-07-30 00:29:20Z
          - now setting dynamic_config => 1 in metadata, as directed in
            CPAN::Meta::Spec

0.003     2013-07-18 17:35:20Z
          - a use of $` has been replaced with ${^PREMATCH}, which requires perl
            5.9.5 (RT#86989)

0.002     2013-07-17 05:31:11Z
          - name, description, and always_recommend options are now prefaced
            with dashes, just like -phase and -relationship (to avoid naming
            conflicts)

0.001     2013-07-17 04:24:19Z
          - Initial release.



Makefile.PL
# This Makefile.PL for Dist-Zilla-Plugin-OptionalFeature was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.016
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.31.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'Module::Build::Tiny' => '0.039',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.009005;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Specify prerequisites for optional features in your distribution",
  "AUTHOR" => "Karen Etheridge <ether\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "Module::Build::Tiny" => "0.039"
  },
  "DISTNAME" => "Dist-Zilla-Plugin-OptionalFeature",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.009005",
  "NAME" => "Dist::Zilla::Plugin::OptionalFeature",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dist::Zilla::Plugin::DynamicPrereqs" => 0,
    "Dist::Zilla::Plugin::MetaResources" => "4.300039",
    "Dist::Zilla::Role::BeforeBuild" => 0,
    "Dist::Zilla::Role::MetaProvider" => 0,
    "Dist::Zilla::Role::PrereqSource" => 0,
    "Module::Runtime" => 0,
    "Moose" => 0,
    "MooseX::Types::Common::String" => 0,
    "MooseX::Types::Moose" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Config::MVP::Reader::INI" => "2.101461",
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "JSON::MaybeXS" => 0,
    "Path::Tiny" => 0,
    "Test::CPAN::Meta::JSON::Version" => 0,
    "Test::CPAN::Meta::YAML::Version" => 0,
    "Test::DZil" => 0,
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "Test::Warnings" => "0.009",
    "YAML::Tiny" => 0,
    "if" => 0,
    "lib" => 0,
    "utf8" => 0
  },
  "VERSION" => "0.020",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Carp" => 0,
  "Config::MVP::Reader::INI" => "2.101461",
  "Dist::Zilla::Plugin::DynamicPrereqs" => 0,
  "Dist::Zilla::Plugin::MetaResources" => "4.300039",
  "Dist::Zilla::Role::BeforeBuild" => 0,
  "Dist::Zilla::Role::MetaProvider" => 0,
  "Dist::Zilla::Role::PrereqSource" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "JSON::MaybeXS" => 0,
  "Module::Build::Tiny" => "0.039",
  "Module::Runtime" => 0,
  "Moose" => 0,
  "MooseX::Types::Common::String" => 0,
  "MooseX::Types::Moose" => 0,
  "Path::Tiny" => 0,
  "Test::CPAN::Meta::JSON::Version" => 0,
  "Test::CPAN::Meta::YAML::Version" => 0,
  "Test::DZil" => 0,
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Test::Warnings" => "0.009",
  "YAML::Tiny" => 0,
  "if" => 0,
  "lib" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


Build.PL
# This Build.PL for Dist-Zilla-Plugin-OptionalFeature was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.014
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.039',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for Dist-Zilla-Plugin-OptionalFeature was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.008.
    use strict;
    use warnings;

    use 5.009005;
    # use Module::Build::Tiny 0.039;
    Module::Build::Tiny::Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.025.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "Module::Build::Tiny" => "0.039",
        "perl" => "5.009005"
      },
      "dist_abstract" => "Specify prerequisites for optional features in your distribution",
      "dist_author" => [
        "Karen Etheridge <ether\@cpan.org>"
      ],
      "dist_name" => "Dist-Zilla-Plugin-OptionalFeature",
      "dist_version" => "0.020",
      "license" => "perl",
      "module_name" => "Dist::Zilla::Plugin::OptionalFeature",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "Carp" => 0,
        "Dist::Zilla::Plugin::DynamicPrereqs" => 0,
        "Dist::Zilla::Plugin::MetaResources" => "4.300039",
        "Dist::Zilla::Role::BeforeBuild" => 0,
        "Dist::Zilla::Role::MetaProvider" => 0,
        "Dist::Zilla::Role::PrereqSource" => 0,
        "Module::Runtime" => 0,
        "Moose" => 0,
        "MooseX::Types::Common::String" => 0,
        "MooseX::Types::Moose" => 0,
        "namespace::autoclean" => 0,
        "perl" => "5.009005",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "Config::MVP::Reader::INI" => "2.101461",
        "Exporter" => 0,
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "JSON::MaybeXS" => 0,
        "Path::Tiny" => 0,
        "Test::CPAN::Meta::JSON::Version" => 0,
        "Test::CPAN::Meta::YAML::Version" => 0,
        "Test::DZil" => 0,
        "Test::Deep" => 0,
        "Test::Fatal" => 0,
        "Test::More" => 0,
        "Test::Warnings" => "0.009",
        "YAML::Tiny" => 0,
        "if" => 0,
        "lib" => 0,
        "perl" => "5.009005",
        "utf8" => 0
      }
    );


    my %fallback_build_requires = (
      "Config::MVP::Reader::INI" => "2.101461",
      "Exporter" => 0,
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "JSON::MaybeXS" => 0,
      "Path::Tiny" => 0,
      "Test::CPAN::Meta::JSON::Version" => 0,
      "Test::CPAN::Meta::YAML::Version" => 0,
      "Test::DZil" => 0,
      "Test::Deep" => 0,
      "Test::Fatal" => 0,
      "Test::More" => 0,
      "Test::Warnings" => "0.009",
      "YAML::Tiny" => 0,
      "if" => 0,
      "lib" => 0,
      "perl" => "5.009005",
      "utf8" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}
